### PR TITLE
User accessible playtime

### PR DESCRIPTION
**Pull request by Titian3**
## About the PR
I have made a Stats window in the character selection menu so people can check how much play time they have in total and per role.

## Why / Balance
Because 10 people asked me to check their play time which is only available via admin command and that was enough for me to say this is bullshit and make it. 
It is also generating 5-10 aHelps a day mainly with people wanting to apply to Salamander.

## Technical details
All client side changes, reusing some existing methods that relate to the role selection and time quota to retrieve the time data. Pretty sure that is only loading up once on connection based on some testing with the new window so it shouldn't be making a bunch of db calls.

## Media
Accessed via the customization menu:
image
Simple window with a table, sortable by role or time.
image

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
None, all new UI and functions.

**Changelog**

:cl: Repo
- add: Stats menu on Character setup for total and role playtime.

### CODE REVIEW

**Comment by lzk228** [comment]
i think in Role Type should be job-name string
and question: Hrs and Min in playtime can be localised?

**Comment by lzk228** [comment]
and why overall shows twice?

**Comment by QuietlyWhisper** [comment]
@metalgearsloth is this kinda what you wanted out of #20517?

**Comment by Titian3** [comment]
> i think in Role Type should be job-name string and question: Hrs and Min in playtime can be localised?

I tried not to touch the serverside to keep it simple so its only bringing back those Job names, ill have a look if maybe i can map them back to the jobname string

The Hrs and Min is localized:
image

**Comment by Titian3** [comment]
> and why overall shows twice?

The overview one and the one in the table?
Mainly ease of viewing for the overview up the top as that was the core issue i was trying to solve. I guess its not really necessary to have it in the table too.

**Comment by Titian3** [comment]
@lzk228 Updated it to be a bit cleaner now, its using the localized job names now too.
image

**Comment by UbaserB** [comment]
Great idea, love the stats button. Only issue is the UI page, it feels really clunky and unorganised. Firstly i feel the role type and role time should be seperated or highlighted some, along with the [XX]Hrs:[XX]Min to be “[XX] Hrs      [XX] Min” or something along those lines. The start of “[XX] Min” in that case would be consistent down the board of all jobs in the y-axis. Man i love UI design lol.

**Comment by Titian3** [comment]
> Great idea, love the stats button. Only issue is the UI page, it feels really clunky and unorganised. Firstly i feel the role type and role time should be seperated or highlighted some, along with the [XX]Hrs:[XX]Min to be “[XX] Hrs [XX] Min” or something along those lines. The start of “[XX] Min” in that case would be consistent down the board of all jobs in the y-axis. Man i love UI design lol.

Like this?
image
Yeah UI stuff has been struggle every time I make a small change even this one, its like 30min to test its working right and rendering correctly...

**Comment by metalgearsloth** [comment]
Add some margins / thickness to it and it'll look a lot better.

**Comment by Titian3** [comment]
> Add some margins / thickness to it and it'll look a lot better.

I have a 10 all round margin on the main window, its a bit clearer on that last screen cap. 
Did you mean thicker for that, or margins on the table bit?

**Comment by metalgearsloth** [comment]
Mainly between each label and its parent container, I'm mainly looking at the left edge here.

**Comment by ficcialfaint** [comment]
Agree with the UI comments above. I'd separate the table's headers (*Role Type* (also why can't it just be *Role* and *Time*?) and *Role Time*). Also I'm pretty sure there is a method (assume in (Content)LocalizationManager) that localizes and generates a pretty time string as the way it is currently displayed is meh

**Comment by Titian3** [comment]
> Mainly between each label and its parent container, I'm mainly looking at the left edge here.

Ok more margins on the child labels in the table:
image

> Agree with the UI comments above. I'd separate the table's headers (_Role Type_ (also why can't it just be _Role_ and _Time_?) and _Role Time_). Also I'm pretty sure there is a method (assume in (Content)LocalizationManager) that localizes and generates a pretty time string because the way it is currently displayed is meh

I've Simplified the headers names, this is all localized so can be set to whatever makes sense on that side.
The sperate the table's headers you mentioned was the Horizontal separator between the header and the table rows I think, that's in there now. 
I specifically used just Hours and minutes because the output from the server/db it is some timespan that is just pain to read. days:hours:minutes:seconds DD:HH:MM:SS. 
No one wants to do some 24x calculation every time they want to know hours in a game. Generally speaking stats for games I've only ever seen in hours, Eg steam tracker is in hours not days and seconds.
image
But if you know the particular time localization i'm happy to check it out to see if it does a better job, I put the RU style hrs and min in this screen cap as a ref with just changing the loc string.

**Comment by lzk228** [comment]
> Ok more margins on the child labels in the table: image

tgat one is great, and good readable

**Comment by notquitehadouken** [comment]
delicious :trollface: 
no more having to annoy admins in ahelp for them to tell me my time hasn't changed since i last played

**Comment by UbaserB** [comment]
Question, is this overall playtime or playtime for that specific character? If the former, i would be interested in being able to see how many hours or what i have played more with those characters :)

**Comment by LankLTE** [comment]
I might be wrong but I don’t think per-character playtime is actually tracked at all

File: Content.Client/Info/PlaytimeStats/PlaytimeStatsWindow.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        if (match.Success &&
+            int.TryParse(match.Groups["hours"].Value, out var hours) &&
+            int.TryParse(match.Groups["minutes"].Value, out var minutes))
+        {
+            return new TimeSpan(hours, minutes, 0);
+        }
+
+        _sawmill.Error($"Couldn't ParsePlaytime string: {playtimeString} Reverting to zero time for entry");
+        return TimeSpan.Zero;
+    }
```
Comment (metalgearsloth) [inline Content.Client/Info/PlaytimeStats/PlaytimeStatsWindow.cs:126]: Sorting by regex parsing is kinda dirty, would be better to just make a private control with the timespan data stored on it instead and use that.

File: Content.Client/Info/PlaytimeStats/PlaytimeStatsWindow.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+
+
+    private void PopulatePlaytimeData()
+    {
+        try
+        {
```
Comment (metalgearsloth) [inline Content.Client/Info/PlaytimeStats/PlaytimeStatsWindow.cs:133]: Don't add the try-catch.

File: Content.Client/Info/PlaytimeStats/PlaytimeStatsWindow.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                OverallPlaytimeLabel.Text = Loc.GetString("ui-playtime-overall", ("time", formattedPlaytime));
+            }
+
+            var rolePlaytimes = _jobRequirementsManager.FetchPlaytimeByRoles();
+
+            RolesPlaytimeList.RemoveAllChildren();
+            PopulatePlaytimeHeader();
+
+            foreach (var rolePlaytime in rolePlaytimes)
+            {
```
Comment (metalgearsloth) [inline Content.Client/Info/PlaytimeStats/PlaytimeStatsWindow.cs:148]: You could just make it IEnumerable and avoid allocating the entire dictionary instead.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by Titian3** [comment]
> Question, is this overall playtime or playtime for that specific character? If the former, i would be interested in being able to see how many hours or what i have played more with those characters :)

Total user playtime, all stats are user based.

**Comment by Titian3** [comment]
@metalgearsloth I maybe should have put all those change requests in separate commits but they are all done now. Thank you for taking some time to review 😊

**Comment by Titian3** [comment]
> I might be wrong but I don’t think per-character playtime is actually tracked at all

I'm going to explicitly mention it is for the user in the window title in the UI, players will probably have the same question for is it user vs character if we are asking it already.
image

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**